h1b/src/hil/flash/driver.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use core::cell::Cell;
+use ::kernel::hil::time::Alarm;
+use super::smart_program::SmartProgramState;
+
+/// The interface between the flash driver and the (real or fake) flash module.
+pub trait HwInterface {
+    /// Returns true if an operation is running, false otherwise.
+    fn is_programming(&self) -> bool;
+
+    /// Read a single word from the flash (non-blocking). offset is in units of
+    /// words and is relative to the start of flash.
+    fn read(&self, offset: usize) -> u32;
+
+    /// Reads the flash error code.
+    fn read_error(&self) -> u16;
+
+    /// Set flash transaction parameters (word offset and size). The word offset
+    /// is relative to the start of flash and the size is in units of words.
+    fn set_transaction(&self, offset: usize, size: usize);
+
+    /// Fill the flash controller's write buffer. data must have a length no
+    /// larger than 32.
+    fn set_write_data(&self, data: &[u32]);
+
+    /// Kick off a smart program execution.
+    fn trigger(&self, opcode: u32);
+}
+
+/// The H1B flash driver. The hardware interface (either the real flash modules
+/// or the fake) is injected to support testing. This will not configure the
+/// globalsec flash regions -- that must be done independently.
+pub struct Driver<'d, A: Alarm + 'd, H: HwInterface + 'd> {
+    alarm: &'d A,
+    client: Cell<Option<&'d Client>>,
+
+    // Hardware interface. Uses shared references rather than mutable references
+    // because the fake interface used in the unit tests is shared with the unit
+    // tests.
+    hw: &'d H,
+
+    // Smart programming state machine, if an operation is ongoing.
+    smart_program_state: Cell<Option<SmartProgramState>>,
+    opcode: Cell<u32>,
+}
+
+/// A client of the Flash driver -- receives callbacks when flash operations
+/// complete.
+pub trait Client {
+    fn erase_done(&self, ::kernel::ReturnCode);
+    fn write_done(&self, ::kernel::ReturnCode);
+}
+
+impl<'d, A: Alarm, H: HwInterface + 'd> Driver<'d, A, H> {
```

Comment:
I'd recommend against the name `Driver` since this is the name of the trait that provides a system call interface.

Reply:
Renamed to `Flash`.